src/debugger.ts

```diff
@@ -0,0 +1,126 @@
+import {
+  debug,
+  window,
+  workspace,
+  DebugAdapterDescriptorFactory,
+  DebugSession,
+  DebugAdapterExecutable,
+  DebugAdapterDescriptor,
+  ExtensionContext,
+  OutputChannel,
+  DebugConfigurationProvider,
+  CancellationToken,
+  DebugConfiguration,
+  ProviderResult,
+} from 'vscode';
+
+import { spawn } from 'child_process';
+import findNargo from './find-nargo';
+import findNearestPackageFrom from './find-nearest-package';
+
+let outputChannel: OutputChannel;
+
+export function activateDebugger(context: ExtensionContext) {
+  outputChannel = window.createOutputChannel('NoirDebugger');
+
+  context.subscriptions.push(
+    debug.registerDebugAdapterDescriptorFactory('noir', new NoirDebugAdapterDescriptorFactory()),
+    debug.registerDebugConfigurationProvider('noir', new NoirDebugConfigurationProvider()),
+    debug.onDidTerminateDebugSession(() => {
+      outputChannel.appendLine("Debug session ended.");
+    }),
+  );
+}
+
+export class NoirDebugAdapterDescriptorFactory implements DebugAdapterDescriptorFactory {
+  async createDebugAdapterDescriptor(
+    _session: DebugSession,
+    _executable: DebugAdapterExecutable,
+  ): ProviderResult<DebugAdapterDescriptor> {
+    const config = workspace.getConfiguration('noir');
+
+    const configuredNargoPath = config.get<string | undefined>('nargoPath');
+    const nargoPath = configuredNargoPath || findNargo();
+
+    return new DebugAdapterExecutable(nargoPath, ['dap']);
+  }
+}
+
+class NoirDebugConfigurationProvider implements DebugConfigurationProvider {
+  async resolveDebugConfiguration(folder: WorkspaceFolder | undefined, config: DebugConfiguration, token?: CancellationToken): ProviderResult<DebugConfiguration> {
+    if (config.program || config.request == 'attach')
+      return config;
```

Comment:
Shouldn't we run preflight checks for user provided configurations as well? I'm not sure how this works, but if `config.program` probably means the configuration comes from the `launch.json` file. I think it makes sense to do a preflight using the provided configuration. Otherwise, this only applies to "new" debugging configurations.